doc/gearbox/gearbox_mdio-HLD.md

```diff
@@ -0,0 +1,246 @@
+
+### Requirements
+
+The syncd docker and daemon use the SAI library to service the NPU programming. The SAI library uses PCIe to access the NPU hardware. The gbsyncd docker and daemon use the PAI library to service the external PHY configuration processing. The PAI library usualy uses MDIO to access the PHY hardware.
+
+It depends on the switch hardware design that the external PHY could be connected to a FPGA or CPLD based MDIO controller or a switch NPU MDIO bus. The FPGA or CPLD based MDIO controller often has linux kernel driver and provides linux sysfs programming interface. The switch NPU MDIO bus uses SAI library hence an Inter-Process-Communication (IPC) mechanism is required between the syncd daemon and gbsyncd daemon.
+![libPAI MDIO access](images/PAI-MDIO.png)
+
+There are 2 mdio access modes: clause 45 and clause 22. Some external PHY uses clause 45 mdio while other external PHY uses clause 22 mdio. The switch NPU sai switch api should distinguish the 2 modes.
+
+When a configured platform target is built, there is only one syncd docker as the SAI library for the configured platform will cover all the device with switch NPU in the same configured platform class. Although the gbsyncd docker is very much similar to the syncd docker, there could be many gbsyncd docker required as there are different PAI library for different PHY and MDIO access method. Our goal is to build a single gbsyncd docker to cover all PHY and MDIO access method.
+
+### Architecture Design 
+
+There are many choices for the IPC mechanism between the syncd daemon and gbsyncd daemon. One performance requirement is that it should finsh firmware download within a reasonable time. Our design choice is to use the Unix socket as the IPC mechanism. Our design has the MDIO IPC server in the syncd daemon with its own thread. A new syncd class MdioIpcServer is added to start a new thread, to create an unix socket, to listen on the socket, to accept connection and to read/reply IPC messages.
+
+There is a corresponding MDIO access IPC client code in the form of dynamic link library which provides the flexiblity to load the library at runtime. Assuming the MDIO access library for sysfs is also in the form of dynamic library, gbsyncd can select the MDIO access library at runtime based on some configuration in the gearbox\_config.json file.
+
+The same gearbox\_config.json file already has the information of the PAI library name. The information can be used to dynamically load the PAI library at runtime.
```

Comment:
Our design is to have a single gbsyncd container. When new PAI/external PHY type is added, only the new library is added to the gbsyncd container, but not to create another gbsyncd container.

Reply:
Each PAI/PHY type may also need specific psai.profile for initialization.

Reply:
When a new platform with external PHY is added to the SONiC image. The sonic device data directories will have the new init PHY config file, gearbox config file and PHY config file in the new "platform" and "hwsku" directories. The new  PAI/PHY driver libraries need to be added to the 'gbsyncd' docker when the SONiC image is built. dlopen()/dlsym() do not require LD_LIBRARY_PATH or ldconfig setup. As long as the new library has a distinguish path/name, it can be open and loaded at runtime.

Reply:
I see dlopen()/dlsym() do not require LD_LIBRARY_PATH.  My point is the library might need its runtime context setup in gbsyncd docker. How to dynamically make it ready?

- In docker init, some shell scripts could be run to help PAI
- How to let each library know the value of SAI_INIT_CONFIG_FILE
- If the library depends on other libs, how to load them automatically in dlopen? Looks they must be under standard lib directory or LD_LIBRARY_PATH?
- The library name, as well as the related script, needs distinguished

Reply:
I understand that the syncd needs a profile name passed by -p option, e.g.
command=/usr/bin/syncd -p /etc/sai.d/pai.profile -x /usr/share/sonic/hwsku/context_config.json -g 1
in the supervisord.conf. The profile has the environment SAI_INIT_CONFIG_FILE which is PHY specific. There are many different ways to solve the issue. Because the changed VendorSai class will parse the gearbox_config.json file and the key "sai_init_config_file": exists in the file, the VendorSai class will override the value of SAI_INIT_CONFIG_FILE from the profile.
The PAI library is packaged as a debian file and the gbsyncd docker build uses the debian package, the PAI library debian package has the dependence built in at package build time. When it is installed in the gbsyncd docker, all dependence package will be installed too.
The PAI library name is in the  gearbox_config.json file  with the key "lib_name": . We can put any related script under /usr/share/sonic/hwsku or platform module which are platform specific.